### DISCUSSION TITLE
How to add type information to macros?

### DISCUSSION THREAD

**Discussion by earltedly** [krzysztofzablocki/SourceryPro-Feedback#5] [Q&A] [answered]
I'm following the workshop tutorial as a way of getting into Sourcery Pro. I've got the AutoCases example...

So I have the code:

```
{% for enum in types.implementing.AutoCases|enum %}
extension {{ enum.name }} {
    static let count = {{ enum.cases.count }}

    {% if not enum.hasAssociatedValues %}
    static let allCases: [{{ enum.name }}] = [
    {% for case in enum.cases %}
        .{{ case.name }}{% if not forloop.last %},{% endif %}
    {% endfor %}]
    {% endif %}
}
{% endfor %}
```

This is generating the correct code. However in the IDE I see a warning on each line where `enum` is used:
<img width="686" alt="image" src="https://user-images.githubusercontent.com/155424/115867832-7c433a00-a433-11eb-8c4c-23da419edc30.png">

`Unknown type enum. cases. count, wrong keypath or needs {% typed %} tag`

The product page says:

> {% typed variable as MyType %} lets you add type information to macros, simply declare it after defining your macro and Sourcery Pro will start providing you with all type information for your variable

I can't seem to figure out the syntax though. I'm guessing it's what's needed to make the errors go away.

**Comment by krzysztofzablocki** [comment, marked as answer]
you need to let Sourcery Pro know what type it's looking at since you used a parent class (Type not Enum, and as such it has no `cases` field) in your for loop, so with `typed` you simply do this:
```swift
{% for enum in types.implementing.AutoCases|enum %}
{% typed enum as Enum %}
extension {{ enum.name }} {
...
```

BTW. when you open a tag you can usually use ctrl+space to get Pro editor to suggest you some code, so if you open `{% typed ` and put your cursor at each keyword placement it will tell you whether to put a unknown variable, `as` keyword or list actual types the tool knows about for you 😉 

in your scenario you can also change your for loop to iterate over enums because those are already collections of `Enum` types so I can infer that it's what you wanted:
```swift
{% for enum in types.enums|implements:"AutoCases" %}
```

**Reply by earltedly** [reply]
That works perfectly - thanks for the fast answer!

```
{% for enum in types.enums|implements:"AutoCases" %}
```
I wondered if there was a way of flipping in too so as to be filtering a typed class rather than something generic. 

Really appreciate you showing both approaches.
